scep/authority.go

```diff
@@ -0,0 +1,511 @@
+	DB nosql.DB
+	// DNS the host used to generate accurate SCEP links. By default the authority
+	// will use the Host from the request, so this value will only be used if
+	// request.Host is empty.
+	DNS string
+	// Prefix is a URL path prefix under which the SCEP api is served. This
+	// prefix is required to generate accurate SCEP links.
+	Prefix string
+}
+
+// SignAuthority is the interface for a signing authority
+type SignAuthority interface {
+	Sign(cr *x509.CertificateRequest, opts provisioner.SignOptions, signOpts ...provisioner.SignOption) ([]*x509.Certificate, error)
+	LoadProvisionerByID(string) (provisioner.Interface, error)
+}
+
+// New returns a new Authority that implements the SCEP interface.
+func New(signAuth SignAuthority, ops AuthorityOptions) (*Authority, error) {
+
+	if _, ok := ops.DB.(*database.SimpleDB); !ok {
+		// If it's not a SimpleDB then go ahead and bootstrap the DB with the
+		// necessary SCEP tables. SimpleDB should ONLY be used for testing.
+		tables := [][]byte{certTable}
+		for _, b := range tables {
+			if err := ops.DB.CreateTable(b); err != nil {
+				return nil, fmt.Errorf("%w: error creating table %s", err, string(b))
+			}
+		}
+	}
+
+	authority := &Authority{
+		backdate: ops.Backdate,
+		db:       ops.DB,
+		prefix:   ops.Prefix,
+		dns:      ops.DNS,
+		signAuth: signAuth,
+	}
+
+	// TODO: this is not really nice to do; the Service should be removed
+	// in its entirety to make this more interoperable with the rest of
+	// step-ca.
+	if ops.Service != nil {
+		authority.intermediateCertificate = ops.Service.certificateChain[0]
+		authority.service = ops.Service
+	}
+
+	return authority, nil
+}
+
+var (
+	// TODO: check the default capabilities; https://tools.ietf.org/html/rfc8894#section-3.5.2
+	defaultCapabilities = []string{
+		"Renewal",
+		"SHA-1",
+		"SHA-256",
+		"AES",
+		"DES3",
+		"SCEPStandard",
+		"POSTPKIOperation",
+	}
+)
+
+// LoadProvisionerByID calls out to the SignAuthority interface to load a
+// provisioner by ID.
+func (a *Authority) LoadProvisionerByID(id string) (provisioner.Interface, error) {
+	return a.signAuth.LoadProvisionerByID(id)
+}
+
+// GetLinkExplicit returns the requested link from the directory.
+func (a *Authority) GetLinkExplicit(provName string, abs bool, baseURL *url.URL, inputs ...string) string {
+	// TODO: taken from ACME; move it to directory (if we need a directory in SCEP)?
+	return a.getLinkExplicit(provName, abs, baseURL, inputs...)
+}
+
+// getLinkExplicit returns an absolute or partial path to the given resource and a base
+// URL dynamically obtained from the request for which the link is being calculated.
+func (a *Authority) getLinkExplicit(provisionerName string, abs bool, baseURL *url.URL, inputs ...string) string {
+
+	// TODO: do we need to provide a way to provide a different suffix/base?
+	// Like "/cgi-bin/pkiclient.exe"? Or would it be enough to have that as the name?
+	link := fmt.Sprintf("/%s", provisionerName)
```

Comment:
It currently is formatted as what you mention here, but the RFC mentions "/cgi-bin/pkiclient.exe" explicitly and I've seen examples of other paths. My comment is about making it possible to configure an optional suffix, so that the full path results in /scep/somename/cgi-bin/pkiclient.exe (or a different suffix).

Reply:
I meant that there's no need to go through all the sprintf parsing for this, totally unnecessary performance improvement, though.